File: examples/mac/dump_tree_axapi.js
#!/usr/bin/env node --preserve-symlinks

const kAttributes = [
  'AXRoleDescription',
  'AXTitle',
  'AXDescription',
  'AXFocused',
];


const {name, pid} = processArguments();

const acacia_axapi = require('./acacia_axapi');
const kSupportedListTypes = new Set([
  acacia_axapi.ValueType_NODE, acacia_axapi.ValueType_STRING,
  acacia_axapi.ValueType_RANGE, acacia_axapi.ValueType_DICTIONARY
]);

const root = getRootAXAPINode(name, pid);
printNodeAndSubtree(root, 0);

function printNodeAndSubtree(node, level) {
  if (!node.hasAttribute('AXRole'))
    return;
  if (node.getValueType('AXRole') == acacia_axapi.ValueType_UNKNOWN)
    return;

  const prefix = '--'.repeat(level) + '>';
  const roleAndAttributes = serializeRoleAndAttributes(node, kAttributes);
  console.log(`${prefix} ${roleAndAttributes}`);

  if (!node.hasAttribute('AXChildren'))
    return;

  let children = getListAttributeValue(node, 'AXChildren');
  for (let child of children)
    printNodeAndSubtree(child, level + 1);
}

function serializeRoleAndAttributes(node, attributes) {
  let output = getAttributeValue(node, 'AXRole');

  for (let attribute of attributes) {
    if (!node.hasAttribute(attribute))
      continue;

    let value = getAttributeValue(node, attribute);
    switch (typeof value) {
      case 'undefined':
        continue;
      case 'boolean':
        if (value)
          output += ` ${attribute}`;
        break;
      case 'string':
        output += ` ${attribute}='${value}'`;
        break;
      default:
        output += ` ${attribute}=${value}`;
        break;
    }
  }

  return output;
}

function getRootAXAPINode(name, pid) {
  if (name !== undefined) {
    let root = acacia_axapi.findRootAXAPINodeForName(name);
    if (root.isNull()) {
      console.error(`No running application found with name ${name}.`)
      process.exit();
    }
    return root;
  }

  if (pid !== undefined) {
    let root = acacia_axapi.findRootAXAPINodeForPID(pid);
    if (!root.isNull()) {
      console.error(`No running application found with PID ${pid}.`)
      process.exit();
    }
    return root;
  }

  printUsageAndExit();
}

function processArguments() {
  let args = process.argv.slice(2);

  if (args.length === 1)
    return {name: args[0]};

  if (args.length === 2) {
    if (args[0] !== '-p')
      printUsageAndExit();
    if (!args[1].match(/\d+/))
      printUsageAndExit();
    return {pid: Number(args[1])};
  }

  printUsageAndExit();
}

function printUsageAndExit() {
  const script_name = process.argv[1];
  console.error(`Usage: ${script_name} [-p <pid> | <name>]`);
  process.exit();
}

function getAttributeValue(node, attribute) {
  try {
    const type = node.getValueType(attribute);
    switch (type) {
      case acacia_axapi.ValueType_LIST:
        return getListAttributeValue(node, attribute);
      case acacia_axapi.ValueType_BOOLEAN:
        return node.getBooleanValue(attribute);
      case acacia_axapi.ValueType_INT:
        return node.getIntValue(attribute);
      case acacia_axapi.ValueType_FLOAT:
        return node.getFloatValue(attribute);
      case acacia_axapi.ValueType_STRING:
        return node.getStringValue(attribute);
      case acacia_axapi.ValueType_URL:
        return node.getURLValue(attribute);
      case acacia_axapi.ValueType_NODE:
        return node.getNodeValue(attribute);
      case acacia_axapi.ValueType_POINT:
        return node.getPointValue(attribute);
      case acacia_axapi.ValueType_SIZE:
        return node.getSizeValue(attribute);
      case acacia_axapi.ValueType_RECT:
        return node.getRectValue(attribute);
      case acacia_axapi.ValueType_RANGE:
        return node.getRangeValue(attribute);
      case acacia_axapi.ValueType_DICTIONARY:
        return node.getDictionaryValue(attribute);
      default:
        const typeString = acacia_axapi.ValueTypeToString(type);
        console.error(`Unsupported type: ${typeString} for ${attribute}`);
        return undefined;
    }
  } catch (e) {
    return `(error: ${e})`;
  }
}

function getListAttributeValue(node, attribute) {
  let count = node.getListElementCount(attribute);
  if (count === 0)
    return [];

  const type = node.getListElementType(attribute);
  if (!kSupportedListTypes.has(type)) {
    const typeString = acacia_axapi.ValueTypeToString(type);
    console.error(`Unsupported list type: ${typeString} for ${attribute}`);
    return undefined;
  }

  let array = [];
  for (let i = 0; i < count; i++)
    array.push(getListAttributeValueAtIndex(node, type, attribute, i));

  return array;
}

function getListAttributeValueAtIndex(node, type, attribute, index) {
  switch (type) {
    case acacia_axapi.ValueType_NODE:
      return node.getNodeListValueAtIndex(attribute, index);
    case acacia_axapi.ValueType_STRING:
      return node.getStringListValueAtIndex(attribute, index);
    case acacia_axapi.ValueType_RANGE:
      return node.getRangeListValueAtIndex(attribute, index);
    case acacia_axapi.ValueType_DICTIONARY:
      return node.getDictionaryListValueAtIndex(attribute, index);
    default:
      return undefined;
  }
}

File: examples/mac/dump_tree_axapi.py
#!/usr/bin/env python3

import argparse
import sys
import acacia_axapi

ATTRIBUTES = [
  'AXRoleDescription',
  'AXTitle',
  'AXDescription',
  'AXFocused',
]


def print_node_and_subtree(node, level):
  prefix = "--" * level
  role_and_attributes = serialize_role_and_attributes(node, ATTRIBUTES)
  print(f'{prefix} {role_and_attributes}')

  if not node.hasAttribute('AXChildren'):
    return

  children = get_list_attribute_value(node, 'AXChildren')
  for child in children:
    print_node_and_subtree(child, level + 1)


def serialize_role_and_attributes(node, attributes):
  output = get_attribute_value(node, 'AXRole')

  for attribute in attributes:
    if not node.hasAttribute(attribute):
      continue

    value = get_attribute_value(node, attribute)
    match type(value).__name__:
      case 'NoneType':
        continue
      case 'bool':
        if value:
          output += f' {attribute}'
      case 'str':
        value_utf16 = value.encode('utf-16', 'surrogatepass').decode('utf-16', 'replace')
        output += f' {attribute}="{value_utf16}"'
      case _:
        output += f' {attribute}={value}'

  return output


def get_attribute_value(node, attribute):
  try:
    type = node.getValueType(attribute)
    match type:
      case acacia_axapi.ValueType_LIST:
        return get_list_attribute_value(node, attribute)
      case acacia_axapi.ValueType_BOOLEAN:
        return node.getBooleanValue(attribute)
      case acacia_axapi.ValueType_INT:
        return node.getIntValue(attribute)
      case acacia_axapi.ValueType_FLOAT:
        return node.getFloatValue(attribute)
      case acacia_axapi.ValueType_STRING:
        return node.getStringValue(attribute)
      case acacia_axapi.ValueType_URL:
        return node.getURLValue(attribute)
      case acacia_axapi.ValueType_NODE:
        return node.getNodeValue(attribute)
      case acacia_axapi.ValueType_POINT:
        return node.getPointValue(attribute)
      case acacia_axapi.ValueType_SIZE:
        return node.getSizeValue(attribute)
      case acacia_axapi.ValueType_RECT:
        return node.getRectValue(attribute)
      case acacia_axapi.ValueType_RANGE:
        return node.getRangeValue(attribute)
      case acacia_axapi.ValueType_DICTIONARY:
        return node.getDictionaryValue(attribute)
      case _:
        type_string = acacia_axapi.ValueTypeToString(type)
        print(f'Unsupported type: {type_string} for {attribute}')
        return None
  except Exception as error:
    return f'(error: {error})'


def get_list_attribute_value(node, attribute):
  count = node.getListElementCount(attribute)
  if count == 0:
    return []

  type = node.getListElementType(attribute)
  list = None
  match type:
    case acacia_axapi.ValueType_NODE:
      list = node.getNodeListValue(attribute)
    case acacia_axapi.ValueType_STRING:
      list = node.getStringListValue(attribute)
    case acacia_axapi.ValueType_RANGE:
      list = node.getRangeListValue(attribute)
    case acacia_axapi.ValueType_DICTIONARY:
      list = node.getDictionaryListValue(attribute)
    case _:
      type_string = acacia_axapi.ValueTypeToString(type)
      sys.stderr.write(f'Unsupported list type: {type_string} for {attribute}.')
      return None

  result = []
  for i in range(count):
    result.append(list[i])
  return result


def process_arguments():
  parser = argparse.ArgumentParser(
      description=
      "Dumps the Mac accessibility tree for the specified application.")
  parser.add_argument("--pid", type=int)
  parser.add_argument("--name", type=str)
  args = parser.parse_args()
  if not (args.pid or args.name):
    parser.print_help()
    sys.exit()
  return args


def main():
  args = process_arguments()
  if args.name:
    print(f'Searching for application with name "{args.name}".')
    root = acacia_axapi.findRootAXAPINodeForName(args.name)
  else:
    print(f'Searching for application with pid {args.pid}.')
    root = acacia_axapi.findRootAXAPINodeForPID(args.pid)

  if root.isNull():
    sys.stderr.write("Error: Application not found.")
    sys.exit()

  try:
    print_node_and_subtree(root, 0)
  except Exception as error:
    sys.stderr.write(f"Error encountered printing node: {error}")


if __name__ == "__main__":
    main()
